bot/cogs/codeblock/parsing.py

```diff
@@ -0,0 +1,226 @@
+            if pattern.match(line):
+                repl_lines += 1
+
+                # Once a pattern is matched, only use that pattern for the remaining lines.
+                patterns = (pattern,)
+                break
+
+        if repl_lines == threshold:
+            log.trace("Content is (I)Python REPL code.")
+            return True
+
+    log.trace("Content is not (I)Python REPL code.")
+    return False
+
+
+def is_python_code(content: str) -> bool:
+    """Return True if `content` is valid Python code or (I)Python REPL output."""
+    dedented = textwrap.dedent(content)
+
+    # Parse AST twice in case _fix_indentation ends up breaking code due to its inaccuracies.
+    return (
+        _is_python_code(dedented)
+        or _is_repl_code(dedented)
+        or _is_python_code(_fix_indentation(content))
+    )
+
+
+def parse_bad_language(content: str) -> Optional[BadLanguage]:
+    """
+    Return information about a poorly formatted Python language in code block `content`.
+
+    If the language is not Python, return None.
+    """
+    log.trace("Parsing bad language.")
+
+    match = _RE_LANGUAGE.match(content)
+    if not match:
+        return None
+
+    return BadLanguage(
+        language=match["lang"],
+        has_leading_spaces=match["spaces"] is not None,
+        has_terminal_newline=match["newline"] is not None,
+    )
+
+
+def _get_leading_spaces(content: str) -> int:
+    """Return the number of spaces at the start of the first line in `content`."""
+    current = content[0]
+    leading_spaces = 0
+
+    while current == " ":
+        leading_spaces += 1
+        current = content[leading_spaces]
+
+    return leading_spaces
+
+
+def _fix_indentation(content: str) -> str:
+    """
+    Attempt to fix badly indented code in `content`.
+
+    In most cases, this works like textwrap.dedent. However, if the first line ends with a colon,
+    all subsequent lines are re-indented to only be one level deep relative to the first line.
+    The intent is to fix cases where the leading spaces of the first line of code were accidentally
+    not copied, which makes the first line appear not indented.
+
+    This is fairly naïve and inaccurate. Therefore, it may break some code that was otherwise valid.
+    It's meant to catch really common cases, so that's acceptable. Its flaws are:
+
+    - It assumes that if the first line ends with a colon, it is the start of an indented block
+    - It uses 4 spaces as the indentation, regardless of what the rest of the code uses
+    """
+    lines = content.splitlines(keepends=True)
+
+    # Dedent the first line
+    first_indent = _get_leading_spaces(content)
+    first_line = lines[0][first_indent:]
+
+    second_indent = _get_leading_spaces(lines[1])
+
+    # If the first line ends with a colon, all successive lines need to be indented one
+    # additional level (assumes an indent width of 4).
+    if first_line.rstrip().endswith(":"):
+        second_indent -= 4
```

Comment:
I'm not quite sure I understand the -4 but this doesn't fix the indentation if it's not copied on the first line for compound statements with multiple clauses like
```py
if ...:
        ...
    elif ...:
        ...
    else:
        ...
```
If we do not care about those but only for definitions (since it's unlikely to happen and probably complex to account for) and it's for decreasing the indent, is it necessary? Python won't care about the width of indentation under one clause.

Correct me here if I misunderstood it; but the multi clause part could at least be added to the docstring as one of the flaws

Reply:
It certainly works fine for me with that example.

Reply:
Seems to work now, guess I mistyped something when testing it.